### ISSUE TITLE
Add Auto function execution

### ISSUE THREAD

**Issue by samratfkt** [openai-php/client#285]
Hello, Is it possible to add auto function execution in this lib? We need to process the function manually now.

**Comment by gehrisandro** [comment]
Hi @samratfkt 

Can you give an example how the API should look like?

**Comment by samratfkt** [comment]
I don't think we can do it via api, but currently I am checking the function call and then using "call_user_func_array($functionName, $argumants)" function to processing the complete calling. So I think if you guys can add the whole function calling into it, it will be more helpful and easy to integrate.

**Comment by yguedidi** [comment]
Maybe https://github.com/theodo-group/LLPhant#function can help here

**Comment by gehrisandro** [comment]
@samratfkt I meant, how you would like to use the package in an ideal world.

@yguedidi Thank you for the link.

A feature to convert PHP functions into a proper request and then execute them when requested is on our todo list.

**Comment by vijaythecoder** [comment]
@hschimpf Love what you did for Function management. I am looking to organize my functions for so long and this is good starter, did you also add support for arrays? or nested parameter structure? would love to see your thoughts on how you approach.

**Comment by samuelgjekic** [comment]
Is the above being implemented already into this library?

**Comment by samuelgjekic** [comment]
Even if i am manually calling functions, how would it look like in php ? I know how to add the functions to the assistant, but how would i be able to make the ai run a php function i declared ? In .net i have done it however with a library built for it.

**Comment by hschimpf** [comment]
Hi @samuelgjekic, I showed an example in my [response above](#issuecomment-1883895076)

> Even if i am manually calling functions, how would it look like in php ? I know how to add the functions to the assistant, but how would i be able to make the ai run a php function i declared ? In .net i have done it however with a library built for it.

See the **Example App** part of my response. After registering the functions and requesting a completion to the chat API, you check if the response has `tool_calls` and just call them using the `call()` method of the example trait.

Hope it helps.

**Comment by vijaythecoder** [comment]
@hschimpf Thanks for getting back, I really appreciate. May be my explanation is bad. Currently, I am trying to achieve the following. I have a tool(function) that has the following schema. 

```php
'items' => [
    'type' => 'object',
    'properties' => [
        'name' => [
            'type' => 'string',
            'description' => 'The name of the item',
        ],
        'variation' => [
            'type' => 'object',
            'properties' => [
                'id' => [
                    'type' => 'string',
                    'description' => 'id of the variation',
                ],
            ],
        ],
    ],
]
```

the part `variations` is an object or a nested parameter for Open AI. With your code, I am not sure how to define it as parameter along with name

```php
public function handle(
        #[Description('The name of the item')]
        string $name,

        #[Description('Variation of the item')]
        string $variation, // But I need this to represent as an object parameter with child items. 
) {
...
```

let me know if you have a work around for that.

**Comment by samuelgjekic** [comment]
> Hi @samuelgjekic, I showed an example in my [response above](#issuecomment-1883895076)
> 
> > Even if i am manually calling functions, how would it look like in php ? I know how to add the functions to the assistant, but how would i be able to make the ai run a php function i declared ? In .net i have done it however with a library built for it.
> 
> See the **Example App** part of my response. After registering the functions and requesting a completion to the chat API, you check if the response has `tool_calls` and just call them using the `call()` method of the example trait.
> 
> Hope it helps.

Thanks, after your reply and after looking into the app example i am starting to understand this a bit better. However i am using assistans api so i would check for tool calls in the thread or am i confused still?

**Comment by hschimpf** [comment]
> the part `variations` is an object or a nested parameter for Open AI. With your code, I am not sure how to define it as parameter along with name
> 
> ```
> public function handle(
>         #[Description('The name of the item')]
>         string $name,
> 
>         #[Description('Variation of the item')]
>         string $variation, // But I need this to represent as an object parameter with child items. 
> ) {
> ...
> ```

@vijaythecoder Oh! I see what you were saying now.

I think that we can tweak the `parseFunctionParameters` to allow the use of classes as parameter type. If the parameters is a class, we can access the properties of it thorugh `ReflectionClass::getProperties()` method, and doing this in a loop will allow support for infinite nested params.

**Comment by hschimpf** [comment]
>> When you initiate a Run with a user Message that triggers the function, the Run will enter a pending status. After it processes, the run will enter a requires_action state which you can verify by retrieving the Run. The model can provide multiple functions to call at once using parallel function calling:
> 
> The above is from the assistant docs on function calling, so maybe its the "requires_action" state that i need to check for ? @hschimpf

Hi @samuelgjekic. Yeah, if you are using the Assistant's API, you should check for `required_action.type = "submit_tool_outputs"`.

You will need to tweak a bit my example, but the general functionality is the same. Check for tool calls, execute them and [submit the results](https://github.com/openai-php/client/?tab=readme-ov-file#submittooloutputs).

**Comment by samuelgjekic** [comment]
> >> When you initiate a Run with a user Message that triggers the function, the Run will enter a pending status. After it processes, the run will enter a requires_action state which you can verify by retrieving the Run. The model can provide multiple functions to call at once using parallel function calling:
> > 
> > The above is from the assistant docs on function calling, so maybe its the "requires_action" state that i need to check for ? @hschimpf
> 
> Hi @samuelgjekic. Yeah, if you are using the Assistant's API, you should check for `required_action.type = "submit_tool_outputs"`.
> 
> You will need to tweak a bit my example, but the general functionality is the same. Check for tool calls, execute them and [submit the results](https://github.com/openai-php/client/?tab=readme-ov-file#submittooloutputs).

Thank you ! I got it working

**Comment by shinevue** [comment]
Wonderful, discussion.
